posts/2019/11/metadsl-talk.md

```diff
@@ -0,0 +1,150 @@
+
+However, to get to those systems it would be good to first make our pattern matching system more restricted. Currently, we can define "pure" rules, which are have strictly structutral matching, and "unpure" rules which execute arbitray Python functions to determine the replacement. The unpure rules will be very hard to reason about mathematically, unless we decide to model all of Python, which seems like a bad idea. So if we can move more, if not all, of our system to pure replacements then we can have more luck analyzing them. I have a couple of idea here:
+
+* One use case for unpure functions is to check if a certain python value has some **property**. Like checking if a integer is 0 to execute some optimization. Even if we don't wanna be able to prove anything about the actual details of these properties, we could model them as pre-checks for the substition based on deterministic functions on the values.
+* Another is to wrap/unwrap from **values in the host language** of Python. We need this at the edges of the system, for example to unwrap tuples or replace based on which python type is found. I don't think we can avoid this part. However, by making sure we limit it to the edges, we could still possibly prove things about the inside of thte system.
+* We also embed a **simply typed lambda calculus** into the language itself, and implement beta reduction by replacing variables in the subtree. ADD LINK FOR THIS. Beta reduction is impure, because we are doing a recursive find a replace on the subtree. Creation, from a Python function, is impure not only because it calls a Python callable, but also because it generates a "Variable" with a unique ID to save as the param for the function.
+* In order to compile a functional data flow expression to an imperative CFG we need to order our expressions (EXAMPLE LINK). One way to do this is to do a preorder (***check) traversal of the graph. More details can be found in XXXX data flow graph paper. It's hard for me at the moment to imagine this kind of transformation in a pure pattern replacement.
+
+These last two have probably been explore in the literature before. It would be good to talk to someone from the pattern matching community to ask about these two.
+
+
+
+### Python Control Flow
+
+
+It was really nice to meet XXXX for the first time. He has a ton of experience hacking on Python to make things faster and was one of the original contributors to the Blaze project at Continiuum.
+
+Currently, `metadsl` creates expressions just by executing Python functions. This is nice and simple, but it doesn't let us map Python control flow, like `if`, `for`, and list comprehensions, to a functional form. XXX said we could modify the CPython interpreter to be able to transform these things. I like this method more than AST or bytecode parsing since it should work without access to the source and we don't need to put our computation in a function. 
```

Comment:
Some automatic AST parsing might be nice, instead of modifying the interpreter.

Reply:
Yeah I was not keen on modifying CPython as well and was looking at AST parsing, before I talked to @llllllllll. His work on [`codetransformer`](https://github.com/llllllllll/codetransformer) looks like it could be useful!